File: test_resources/kotlin/enum_extended.kt
sealed interface MyEnum {
    data object Idle : MyEnum

    data class Move(
        val x: Double,
        val y: Double,
    ) : MyEnum

    data class Update(
        val p1: Double,
        val p2: Double,
        val p4: String,
    ) : MyEnum

    companion object {
        fun createDefault(): MyEnum = Idle

        fun readFromBuffers(reader: Long): MyEnum {
            return when (val case = UInt.readFromBuffers(reader)) {
                1U -> Idle
                2U -> {
                    val x = Double.readFromBuffers(reader)
                    val y = Double.readFromBuffers(reader)

                    Move(
                        x = x,
                        y = y,
                    )
                }
                3U -> {
                    val p1 = Double.readFromBuffers(reader)
                    val p2 = Double.readFromBuffers(reader)
                    val p4 = String.readFromBuffers(reader)

                    Update(
                        p1 = p1,
                        p2 = p2,
                        p4 = p4,
                    )
                }
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            Idle -> {}
            is Move -> {
                x.writeToBuffers(writer)
                y.writeToBuffers(writer)
            }
            is Update -> {
                p1.writeToBuffers(writer)
                p2.writeToBuffers(writer)
                p4.writeToBuffers(writer)
            }
        }
    }
}

sealed interface MyEnumWithoutPositions {
    data class Option1(
        val p0: ULong,
    ) : MyEnumWithoutPositions

    data class Option2(
        val name: String,
    ) : MyEnumWithoutPositions

    data object Option3 : MyEnumWithoutPositions

    data object Option4 : MyEnumWithoutPositions

    companion object {
        fun createDefault(): MyEnumWithoutPositions = Option1(
            0UL,
        )

        fun readFromBuffers(reader: Long): MyEnumWithoutPositions {
            return when (val case = UInt.readFromBuffers(reader)) {
                0U -> {
                    val p0 = ULong.readFromBuffers(reader)

                    Option1(
                        p0 = p0,
                    )
                }
                1U -> {
                    val name = String.readFromBuffers(reader)

                    Option2(
                        name = name,
                    )
                }
                2U -> Option3
                3U -> Option4
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            is Option1 -> {
                p0.writeToBuffers(writer)
            }
            is Option2 -> {
                name.writeToBuffers(writer)
            }
            Option3 -> {}
            Option4 -> {}
        }
    }
}

sealed interface MyEnumWithManyArguments {
    data class Option1(
        val p0: ULong,
        val p1: ULong,
        val p2: ULong,
    ) : MyEnumWithManyArguments

    data class Option2(
        val name: String,
    ) : MyEnumWithManyArguments

    data object Option3 : MyEnumWithManyArguments

    data object Option4 : MyEnumWithManyArguments

    companion object {
        fun createDefault(): MyEnumWithManyArguments = Option1(
            0UL,
            0UL,
            0UL,
        )

        fun readFromBuffers(reader: Long): MyEnumWithManyArguments {
            return when (val case = UInt.readFromBuffers(reader)) {
                0U -> {
                    val p0 = ULong.readFromBuffers(reader)
                    val p1 = ULong.readFromBuffers(reader)
                    val p2 = ULong.readFromBuffers(reader)

                    Option1(
                        p0 = p0,
                        p1 = p1,
                        p2 = p2,
                    )
                }
                1U -> {
                    val name = String.readFromBuffers(reader)

                    Option2(
                        name = name,
                    )
                }
                2U -> Option3
                3U -> Option4
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            is Option1 -> {
                p0.writeToBuffers(writer)
                p1.writeToBuffers(writer)
                p2.writeToBuffers(writer)
            }
            is Option2 -> {
                name.writeToBuffers(writer)
            }
            Option3 -> {}
            Option4 -> {}
        }
    }
}

sealed interface MyEnumWithNamedArguments {
    data class Option1(
        val name: String,
        val value: Float,
        val bytes: List<UByte>,
    ) : MyEnumWithNamedArguments

    data class Option2(
        val p0: ULong,
        val p1: ULong,
        val p2: ULong,
    ) : MyEnumWithNamedArguments

    data object Option3 : MyEnumWithNamedArguments

    data object Option4 : MyEnumWithNamedArguments

    companion object {
        fun createDefault(): MyEnumWithNamedArguments = Option1(
            name = "",
            value = 0f,
            bytes = listOf(),
        )

        fun readFromBuffers(reader: Long): MyEnumWithNamedArguments {
            return when (val case = UInt.readFromBuffers(reader)) {
                0U -> {
                    val name = String.readFromBuffers(reader)
                    val value = Float.readFromBuffers(reader)
                    val bytes = readFromBuffersList(reader) {
                        UByte.readFromBuffers(reader)
                    }

                    Option1(
                        name = name,
                        value = value,
                        bytes = bytes,
                    )
                }
                1U -> {
                    val p0 = ULong.readFromBuffers(reader)
                    val p1 = ULong.readFromBuffers(reader)
                    val p2 = ULong.readFromBuffers(reader)

                    Option2(
                        p0 = p0,
                        p1 = p1,
                        p2 = p2,
                    )
                }
                2U -> Option3
                3U -> Option4
                else -> throw IllegalArgumentException("Invalid enum value: $case")
            }
        }

        fun skipInBuffers(reader: Long, count: Int) {
            for (i in 0..<count) {
                readFromBuffers(reader)
            }
        }
    }

    fun writeToBuffers(writer: Long) {
        when (this) {
            is Option1 -> {
                name.writeToBuffers(writer)
                value.writeToBuffers(writer)
                writeToBuffersList(writer, bytes) { bytesItem ->
                    bytesItem.writeToBuffers(writer)
                }
            }
            is Option2 -> {
                p0.writeToBuffers(writer)
                p1.writeToBuffers(writer)
                p2.writeToBuffers(writer)
            }
            Option3 -> {}
            Option4 -> {}
        }
    }
}

File: test_resources/swift/enum_extended.swift
enum MyEnum: TechPawsBuffersModel {
    case idle
    case move(
        /* x */ Double,
        /* y */ Double
    )
    case update(
        Double,
        Double,
        String
    )

    static func createBuffersDefault() -> MyEnum {
        return .idle
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 1:
            return .idle
        case 2:
            let x = bytesReader.readDouble()
            let y = bytesReader.readDouble()

            return .move(
                x,
                y
            )
        case 3:
            let p1 = bytesReader.readDouble()
            let p2 = bytesReader.readDouble()
            let p4 = String.readFromBuffers(bytesReader)

            return .update(
                p1,
                p2,
                p4
            )
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 1:
                continue
            case 2:
                let _ = bytesReader.readDouble()
                let _ = bytesReader.readDouble()
            case 3:
                let _ = bytesReader.readDouble()
                let _ = bytesReader.readDouble()
                let _ = String.readFromBuffers(bytesReader)
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .idle:
            bytesWriter.writeUInt32(1)
        case .move(
            let x,
            let y
        ):
            bytesWriter.writeUInt32(2)
            bytesWriter.writeDouble(x)
            bytesWriter.writeDouble(y)
        case .update(
            let p1,
            let p2,
            let p4
        ):
            bytesWriter.writeUInt32(3)
            bytesWriter.writeDouble(p1)
            bytesWriter.writeDouble(p2)
            p4.writeToBuffers(bytesWriter)
        }
    }
}

enum MyEnumWithoutPositions: TechPawsBuffersModel {
    case option1(
        UInt64
    )
    case option2(
        /* name */ String
    )
    case option3
    case option4

    static func createBuffersDefault() -> MyEnumWithoutPositions {
        return .option1(
            0
        )
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 0:
            let p0 = bytesReader.readUInt64()

            return .option1(
                p0
            )
        case 1:
            let name = String.readFromBuffers(bytesReader)

            return .option2(
                name
            )
        case 2:
            return .option3
        case 3:
            return .option4
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 0:
                let _ = bytesReader.readUInt64()
            case 1:
                let _ = String.readFromBuffers(bytesReader)
            case 2:
                continue
            case 3:
                continue
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .option1(
            let p0
        ):
            bytesWriter.writeUInt32(0)
            bytesWriter.writeUInt64(p0)
        case .option2(
            let name
        ):
            bytesWriter.writeUInt32(1)
            name.writeToBuffers(bytesWriter)
        case .option3:
            bytesWriter.writeUInt32(2)
        case .option4:
            bytesWriter.writeUInt32(3)
        }
    }
}

enum MyEnumWithManyArguments: TechPawsBuffersModel {
    case option1(
        UInt64,
        UInt64,
        UInt64
    )
    case option2(
        /* name */ String
    )
    case option3
    case option4

    static func createBuffersDefault() -> MyEnumWithManyArguments {
        return .option1(
            0,
            0,
            0
        )
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 0:
            let p0 = bytesReader.readUInt64()
            let p1 = bytesReader.readUInt64()
            let p2 = bytesReader.readUInt64()

            return .option1(
                p0,
                p1,
                p2
            )
        case 1:
            let name = String.readFromBuffers(bytesReader)

            return .option2(
                name
            )
        case 2:
            return .option3
        case 3:
            return .option4
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 0:
                let _ = bytesReader.readUInt64()
                let _ = bytesReader.readUInt64()
                let _ = bytesReader.readUInt64()
            case 1:
                let _ = String.readFromBuffers(bytesReader)
            case 2:
                continue
            case 3:
                continue
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .option1(
            let p0,
            let p1,
            let p2
        ):
            bytesWriter.writeUInt32(0)
            bytesWriter.writeUInt64(p0)
            bytesWriter.writeUInt64(p1)
            bytesWriter.writeUInt64(p2)
        case .option2(
            let name
        ):
            bytesWriter.writeUInt32(1)
            name.writeToBuffers(bytesWriter)
        case .option3:
            bytesWriter.writeUInt32(2)
        case .option4:
            bytesWriter.writeUInt32(3)
        }
    }
}

enum MyEnumWithNamedArguments: TechPawsBuffersModel {
    case option1(
        /* name */ String,
        /* value */ Float,
        /* bytes */ [UInt8]
    )
    case option2(
        UInt64,
        UInt64,
        UInt64
    )
    case option3
    case option4

    static func createBuffersDefault() -> MyEnumWithNamedArguments {
        return .option1(
            /* name */ "",
            /* value */ 0,
            /* bytes */ []
        )
    }

    static func readFromBuffers(
        _ bytesReader: TechPawsBuffersBytesReader
    ) -> Self {
        let caseValue = bytesReader.readUInt32()

        switch caseValue {
        case 0:
            let name = String.readFromBuffers(bytesReader)
            let value = bytesReader.readFloat()
            let bytes = [UInt8].readFromBuffers(bytesReader)

            return .option1(
                name,
                value,
                bytes
            )
        case 1:
            let p0 = bytesReader.readUInt64()
            let p1 = bytesReader.readUInt64()
            let p2 = bytesReader.readUInt64()

            return .option2(
                p0,
                p1,
                p2
            )
        case 2:
            return .option3
        case 3:
            return .option4
        default:
            fatalError("Invalid value: \(caseValue)")
        }
    }

    static func skipInBuffers(
        _ bytesReader: TechPawsBuffersBytesReader,
        _ count: UInt64
    ) {
        for _ in 1...count {
            let caseValue = bytesReader.readUInt32()

            switch caseValue {
            case 0:
                let _ = String.readFromBuffers(bytesReader)
                let _ = bytesReader.readFloat()
                let _ = [UInt8].readFromBuffers(bytesReader)
            case 1:
                let _ = bytesReader.readUInt64()
                let _ = bytesReader.readUInt64()
                let _ = bytesReader.readUInt64()
            case 2:
                continue
            case 3:
                continue
            default:
                fatalError("Invalid value: \(caseValue)")
            }
        }
    }

    func writeToBuffers(
        _ bytesWriter: TechPawsBuffersBytesWriter
    ) {

        switch self {
        case .option1(
            let name,
            let value,
            let bytes
        ):
            bytesWriter.writeUInt32(0)
            name.writeToBuffers(bytesWriter)
            bytesWriter.writeFloat(value)
            bytes.writeToBuffers(bytesWriter)
        case .option2(
            let p0,
            let p1,
            let p2
        ):
            bytesWriter.writeUInt32(1)
            bytesWriter.writeUInt64(p0)
            bytesWriter.writeUInt64(p1)
            bytesWriter.writeUInt64(p2)
        case .option3:
            bytesWriter.writeUInt32(2)
        case .option4:
            bytesWriter.writeUInt32(3)
        }
    }
}
